refactor(semantic): store all data in `PostTransformChecker` in transform checker

overlookmotel opened the pull request:
Pure refactor of transform checker. Store all scope data in `PostTransformChecker` so it doesn't need to be passed around. `SemanticData` contains a full set of all semantic data for semantic pass, so we have 2 instances of it for 1. after transform and 2. rebuilt semantic.

overlookmotel commented:
* **#5050** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5050?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
This is last bit of refactoring. Now I'll start implementing more checks.

Boshen reviewed (commented):
You know how I never use top level functions as entry points 😄

overlookmotel committed:
refactor(semantic): store all data in `PostTransformChecker` in transform checker (#5050)

Pure refactor of transform checker. Store all scope data in `PostTransformChecker` so it doesn't need to be passed around. `SemanticData` contains a full set of all semantic data for semantic pass, so we have 2 instances of it for 1. after transform and 2. rebuilt semantic.

overlookmotel commented:
> You know how I never use top level functions as entry points 😄

Yes, but I was never sure if that was a solid preference or just "how it is". Would you prefer this not to be a top level function?
